[question] Conan + Meson + sysroot issue

### What is your question?

Hi!
I'm trying to build Qt with a prebuilt GCC compiler, having an external sysroot containing libc and relevant libraries (pthread, libm, libdl, etc), meaning that the toolchain itself does not have these libraries inside.

So I defined a toolchain profile with a sysroot:

```
...
[buildenv]
PATH=+(path){{toolchain_path}}/bin
LD_LIBRARY_PATH=+(path){{toolchain_path}}/lib
C_INCLUDE_PATH=+(path){{sysroot}}/usr/include
CC={{toolchain_path}}/bin/{{cc_compiler}}
CXX={{toolchain_path}}/bin/{{cxx_compiler}}
LD={{toolchain_path}}/bin/ld
AR={{toolchain_path}}/bin/ar
AS={{toolchain_path}}/bin/as
NM={{toolchain_path}}/bin/nm
RANLIB={{toolchain_path}}/bin/ranlib
STRIP={{toolchain_path}}/bin/strip
LDFLAGS=-Wl,-rpath-link {{sysroot}}/lib

[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=9
os=Linux

[conf]
tools.build:sysroot={{sysroot}}
tools.build:sharedlinkflags=["-Wl,-rpath-link {{sysroot}}/lib"]
```

So far every dependency built successfully, except freetype:
```
[2/4] Compiling C object libfreetype.a.p/src_bzip2_ftbzip2.c.o
FAILED: libfreetype.a.p/src_bzip2_ftbzip2.c.o
<toolchain_path>/gcc-9.3.0-x86_64/bin/gcc -Ilibfreetype.a.p -I. -I../src -I../src/include -I<sysroot_path>/<home>/conan/home/p/b/zlib337c8139e577f/p/include
...
 -MD -MQ libfreetype.a.p/src_bzip2_ftbzip2.c.o -MF libfreetype.a.p/src_bzip2_ftbzip2.c.o.d -o libfreetype.a.p/src_bzip2_ftbzip2.c.o -c ../src/src/bzip2/ftbzip2.c
../src/src/bzip2/ftbzip2.c:46:10: fatal error: bzlib.h: No such file or directory
```
The problem is this line (replaced the hard-coded paths with a reference):
```
-I<sysroot_path>/<home>/conan/home/p/b/zlib337c8139e577f/p/include
```
The sysroot path is directly prepended to all dependent includes, zlib included, making the zlib header lookup messed up.
I've done some research and found a relevant [official meson issue](https://github.com/mesonbuild/meson/issues/8448). It says:

> Now the issues is, that pkgconf PREFIXES every path in the *.pc files inside dependencies with the sysroot defined in sys_root. However the paths defined pc files in the folder dependencies MUST not be prefixed. Else the #include and linking will fail.

Here's the pc file for zlib:
`build-release/conan/zlib.pc`:
```
prefix=<home>/conan/home/p/b/zlib337c8139e577f/p
libdir=${prefix}/lib
includedir=${prefix}/include
bindir=${prefix}/bin

Name: zlib
Description: Conan package: zlib
Version: 1.3.1
Libs: -L"${libdir}" -lz
Cflags: -I"${includedir}"
```

I've tried nullifying the `PKG_CONFIG_SYSROOT_DIR` variable via [buildenv] in my profile to force pkgconfig to disregard it, to no avail.
Any ideas? This is a very basic cross-compile issue for conan -> meson, it should be straightforward.

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

> A way to work around this could be:
> 
>     * Add an extra machine file to override the `sys_root` Meson property:
> 
> 
> ```shell
> $ conan [YOUR COMMAND] -c "tools.meson.mesontoolchain:extra_machine_files=['/path/to/file/sys_root_empty.ini']"
> ```
> 
> Where that `sys_root_empty.ini` looks like:
> 
> ```
> [properties]
> sys_root = ''
> ```
> 
> Let me know if it worked or if we need to look into it more.

This doesn't work (anymore) for me. Not even when I add the following line to  the recipe of the cross-compiled package (fontconfig in my case):

`tc.properties["sys_root"] = ""`

With this line the sys_root property in the generated cross-file becomes empty but still the sys-root is prepended to all include paths:

```
freetype/2.13.2: Meson configure cmd: meson setup --cross-file "/home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/build-release/conan/conan_meson_cross.ini" --cross-file "/home/wdobbe/.conan2/meson_sys_root_empty.ini" "/home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/build-release" "/home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/src" --prefix=/
freetype/2.13.2: RUN: meson setup --cross-file "/home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/build-release/conan/conan_meson_cross.ini" --cross-file "/home/wdobbe/.conan2/meson_sys_root_empty.ini" "/home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/build-release" "/home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/src" --prefix=/
WARNING: Unknown CPU family cortexa9hf-neon, please report this at https://github.com/mesonbuild/meson/issues/new
The Meson build system
Version: 1.3.2
Source dir: /home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/src
Build dir: /home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/build-release
Build type: cross build
Project name: freetype2
Project version: 2.13.2
C compiler for the host machine: arm-poky-linux-gnueabi-gcc (gcc 9.2.0 "arm-poky-linux-gnueabi-gcc (GCC) 9.2.0")
C linker for the host machine: arm-poky-linux-gnueabi-gcc ld.bfd 2.32.0.20190204
C compiler for the build machine: cc (gcc 7.5.0 "cc (SUSE Linux) 7.5.0")
C linker for the build machine: cc ld.bfd 2.41.0.20230908-150100
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: cortexa9hf-neon
Host machine cpu: cortexa9hf-neon
Target machine cpu family: cortexa9hf-neon
Target machine cpu: cortexa9hf-neon
Program python3 found: YES (/usr/local/python39env/bin/python3)
Has header "unistd.h" : YES 
Has header "fcntl.h" : YES 
Has header "sys/mman.h" : YES 
Found pkg-config: YES (/home/wdobbe/.conan2/p/b/pkgco6b2707c42a451/p/bin/pkgconf) 2.1.0
Run-time dependency zlib found: YES 1.2.13
Run-time dependency bzip2 found: YES 1.0.8
Run-time dependency libpng found: YES 1.6.42
Dependency harfbuzz skipped: feature harfbuzz disabled
Run-time dependency libbrotlidec found: YES 1.1.0
Build targets in project: 5

freetype2 2.13.2

  Operating System
    OS         : linux

  Used Libraries
    Zlib       : system
    Bzip2      : yes
    Png        : yes
    Harfbuzz   : no
    Brotli     : yes

  User defined options
    Cross files: /home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/build-release/conan/conan_meson_cross.ini
                 /home/wdobbe/.conan2/meson_sys_root_empty.ini
    prefix     : /

Found ninja-1.12.1 at /home/wdobbe/.conan2/p/b/ninjabf76fb6402bba/p/bin/ninja
                                                                                                                                                                                                                  
freetype/2.13.2: Meson build cmd: meson compile -C "/home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/build-release" -j64
freetype/2.13.2: RUN: meson compile -C "/home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/build-release" -j64
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /home/wdobbe/.conan2/p/b/ninjabf76fb6402bba/p/bin/ninja -j 64
[18/46] Compiling C object libfreetype.a.p/src_sfnt_sfnt.c.o
FAILED: libfreetype.a.p/src_sfnt_sfnt.c.o 
arm-poky-linux-gnueabi-gcc -Ilibfreetype.a.p -I. -I../src -I../src/include -I/home/wdobbe/.conan2/p/b/swpt_fb08c05e04578/p/sysroots/cortexa9hf-neon-poky-linux-gnueabi/home/wdobbe/.conan2/p/b/zlib7dbebab8830af/p/include -I/home/wdobbe/.conan2/p/b/swpt_fb08c05e04578/p/sysroots/cortexa9hf-neon-poky-linux-gnueabi/home/wdobbe/.conan2/p/b/bzip2e92281329ac89/p/include -I/home/wdobbe/.conan2/p/b/swpt_fb08c05e04578/p/sysroots/cortexa9hf-neon-poky-linux-gnueabi/home/wdobbe/.conan2/p/b/libpne0671a1ce0e00/p/include -I/home/wdobbe/.conan2/p/b/swpt_fb08c05e04578/p/sysroots/cortexa9hf-neon-poky-linux-gnueabi/home/wdobbe/.conan2/p/b/brotl76641bb1988e1/p/include -I/home/wdobbe/.conan2/p/b/swpt_fb08c05e04578/p/sysroots/cortexa9hf-neon-poky-linux-gnueabi/home/wdobbe/.conan2/p/b/brotl76641bb1988e1/p/include/brotli -fvisibility=hidden -fdiagnostics-color=always -DNDEBUG -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -marm -mfpu=neon -mfloat-abi=hard -mcpu=cortex-a9 -fstack-protector-strong -Wformat -Wformat-security -Werror=format-security --sysroot=/home/wdobbe/.conan2/p/b/swpt_fb08c05e04578/p/sysroots/cortexa9hf-neon-poky-linux-gnueabi -O2 -pipe -feliminate-unused-debug-types -marm -mfpu=neon -mfloat-abi=hard -mcpu=cortex-a9 -fstack-protector-strong -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/home/wdobbe/.conan2/p/b/swpt_fb08c05e04578/p/sysroots/cortexa9hf-neon-poky-linux-gnueabi -O2 -feliminate-unused-debug-types --sysroot=/home/wdobbe/.conan2/p/b/swpt_fb08c05e04578/p/sysroots/cortexa9hf-neon-poky-linux-gnueabi -fPIC '-DFT_CONFIG_MODULES_H=<ftmodule.h>' -DFT2_BUILD_LIBRARY=1 '-DFT_CONFIG_OPTIONS_H=<ftoption.h>' '-DFT_CONFIG_CONFIG_H=<ftconfig.h>' -MD -MQ libfreetype.a.p/src_sfnt_sfnt.c.o -MF libfreetype.a.p/src_sfnt_sfnt.c.o.d -o libfreetype.a.p/src_sfnt_sfnt.c.o -c ../src/src/sfnt/sfnt.c
In file included from ../src/src/sfnt/sfnt.c:25:
../src/src/sfnt/sfwoff2.c:27:10: fatal error: brotli/decode.h: No such file or directory
   27 | #include <brotli/decode.h>
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
[45/46] Compiling C object libfreetype.a.p/src_truetype_truetype.c.o
ninja: build stopped: subcommand failed.

freetype/2.13.2: ERROR: 
Package 'a801309751333469de9145f28887e6fc7fe500a8' build failed
freetype/2.13.2: WARN: Build folder /home/wdobbe/.conan2/p/b/freet6b9a8b9ee6f3a/b/build-release
ERROR: freetype/2.13.2: Error in build() method, line 114
        meson.build()
        ConanException: Error 1 while executing
```

--- a/conan/tools/meson/toolchain.py
+++ b/conan/tools/meson/toolchain.py
@@ -289,8 +289,6 @@ class MesonToolchain:
 
         # Read configuration for sys_root property (honoring existing conf)
         self._sys_root = self._conanfile_conf.get("tools.build:sysroot", check_type=str)
-        if self._sys_root:
-            self.properties["sys_root"] = self._sys_root
 
         # Read configuration for compilers
         compilers_by_conf = self._conanfile_conf.get("tools.build:compiler_executables", default={},

--- a/test/integration/toolchains/meson/test_mesontoolchain.py
+++ b/test/integration/toolchains/meson/test_mesontoolchain.py
@@ -690,8 +690,8 @@ def test_compiler_path_with_spaces():
 def test_meson_sysroot_app():
     """Testing when users pass tools.build:sysroot on the profile with Meson
 
-    The generated conan_meson_cross.ini needs to contain both sys_root property to fill the
-    PKG_CONFIG_PATH and the compiler flags with --sysroot.
+    * The generated conan_meson_cross.ini does not fill the "sys_root" property (see https://github.com/conan-io/conan/issues/16468)
+    * It adds the compiler flags with --sysroot.
 
     When cross-building, Meson needs both compiler_executables in the config, otherwise it will fail
     when running setup.
@@ -728,7 +728,7 @@ def test_meson_sysroot_app():
     client.run("install . -pr:h host -pr:b build")
     # Check the meson configuration file
     conan_meson = client.load("conan_meson_cross.ini")
-    assert f"sys_root = '{sysroot}'\n" in conan_meson
+    assert f"sys_root = '{sysroot}'\n" not in conan_meson
     assert re.search(r"c_args =.+--sysroot={}.+".format(sysroot), conan_meson)
     assert re.search(r"c_link_args =.+--sysroot={}.+".format(sysroot), conan_meson)
     assert re.search(r"cpp_args =.+--sysroot={}.+".format(sysroot), conan_meson)

